posts/2020/crud-pattern-in-react-implementation.md

```diff
@@ -0,0 +1,130 @@
+
+## Inverse Data Flow
+
+- A callback function (`openForm()` or `delete()`) defined in the parent (`Books`) component is passed as props to the child (`BookTable`) component.
+- The function is invoked by the child component for an event .
+- eg. to trigger update or delete action (function invoke) on button click (event).
+- The child component can pass data (book id or book object) as an argument to the callback function. As the callback function is defined in the parent, the data is made available to the parent for further action.
+- As data is passed in the _inverse_ direction (from child to parent), Inverse Data Flow is achieved.
+- The passed data can then be used by the parent component to update internal state which may potentially re-render the page.
+
+# BookForm Component - Implementation
+
+A form component creates or updates an entity on the server by performing API actions. In order to do this, it needs to fulfill following requirements:
+
+- Validate form and display validation errors (if any) of specific field
+- Display API response errors (if any) of specific field on the UI
+
+We can fulfill these requirements by using a `formFieldAttributes` object that defines validation rules for each field. We will check its details in the next section.
+
+## `formFieldAttributes` object
+
+It defines custom validation rules for every field in `{fieldName: { attr1: value1, ..., attrN: valueN }}` format. The rules can be specified as attributes as following :
+
+1. `key`: Used to generate and map error to the respective field
+2. `required`: Defines whether an empty field is allowed or not
+3. `customValidator()`: Provides custom validation logic for each field. Returns an error object in `{key: { error: true, helperText:"text" }}` format.
+
+```javascript
+const formFieldAttributes = {
+  title: {          // value of key ("title") is same as this (convention)
+    key: "title",   // Should be unique as it is used as field 'id'
+    required: true, // Whether an empty field is allowed or not
+    customValidator: value =>
+      value.length > 100 // Custom validation logic
+        ? { helperText: "enter less than 100 characters", error: true } // Object with error
+        : { helperText: "", error: false } // Object without error
+  },
+  ...
+}
+
+// Using 'key' as argument to fieldError() and fieldHelperText()
+// to map error and helperText
+<form id="bookForm">
+    <TextField
+        id={formFieldAttributes.title.key}
+        error={this.fieldError(formFieldAttributes.title.key)}
+        helperText={this.fieldHelperText(formFieldAttributes.title.key)}
+        ...
+    />
+...
+</form>
+
+```
+
+`formValidator()` function uses this object to validate form and generate errors (if any). We will check its functioning in the next section.
+
+## Form Validation (`formValidator()` function)
+
```

Comment:
Let's add this paragraph before the code

`formValidator` function is the central place for performing all the validation before API call is being made. Results of validation are collected in a `formErrors` object and then  errors if any present in the `formErrors` object are mapped to the UI by the caller. To perform this, the function uses the `formFieldAttributes` object described above.